pkg/operator/controller/controller.go

```diff
@@ -284,90 +332,71 @@ func (r *reconciler) ensureRouterNamespace() error {
 	return nil
 }
 
-// ensureRouterForIngress ensures all necessary router resources exist for a
+// ensureClusterIngress ensures all necessary router resources exist for a
 // given clusteringress.
-func (r *reconciler) ensureRouterForIngress(ci *ingressv1alpha1.ClusterIngress, caSecret *corev1.Secret, infraConfig *configv1.Infrastructure, dnsConfig *configv1.DNS, result *reconcile.Result) error {
-	expected, err := r.ManifestFactory.RouterDeployment(ci)
-	if err != nil {
-		return fmt.Errorf("failed to build router deployment: %v", err)
-	}
-	current := expected.DeepCopy()
-	err = r.Client.Get(context.TODO(), types.NamespacedName{Namespace: expected.Namespace, Name: expected.Name}, current)
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return fmt.Errorf("failed to get router deployment %s/%s, %v", expected.Namespace, expected.Name, err)
-		}
-
-		err = r.Client.Create(context.TODO(), current)
-		if err == nil {
-			logrus.Infof("created router deployment %s/%s", current.Namespace, current.Name)
-		} else if !errors.IsAlreadyExists(err) {
-			return fmt.Errorf("failed to create router deployment %s/%s: %v", current.Namespace, current.Name, err)
-		}
-	}
+func (r *reconciler) ensureClusterIngress(ci *ingressv1alpha1.ClusterIngress, ingressDomain string, caSecret *corev1.Secret, infraConfig *configv1.Infrastructure, dnsConfig *configv1.DNS, result *reconcile.Result) error {
+	errs := []error{}
 
-	if changed, updated := deploymentConfigChanged(current, expected); changed {
-		err = r.Client.Update(context.TODO(), updated)
-		if err == nil {
-			logrus.Infof("updated router deployment %s/%s", updated.Namespace, updated.Name)
-			current = updated
-		} else {
-			return fmt.Errorf("failed to update router deployment %s/%s, %v", updated.Namespace, updated.Name, err)
-		}
-	}
+	ha := effectiveHighAvailability(ci, infraConfig)
 
-	trueVar := true
-	deploymentRef := metav1.OwnerReference{
-		APIVersion: "apps/v1",
-		Kind:       "Deployment",
-		Name:       current.Name,
-		UID:        current.UID,
-		Controller: &trueVar,
-	}
+	deployment, err := r.ensureRouterDeployment(ci, ingressDomain, infraConfig, ha)
+	errs = append(errs, err)
```

Comment:
`if err != nil`?

Reply:
`utilerrors.NewAggregate` will ignore nil values, but I can add the check if you prefer.

Reply:
Seems weird to keep a nil entry in the array which is a plain `[]error`, error prone if you were to check the len of the array or feed to a function that has different assumptions... err on the side of consistency?

Reply:
Fixed.